js/buda.js

```diff
@@ -354,43 +354,65 @@ module.exports = class buda extends Exchange {
             'market': market['id'],
         };
         const response = await this.publicGetMarketsMarketTicker (this.extend (request, params));
+        //
+        //     {
+        //         "ticker":{
+        //             "market_id":"ETH-BTC",
+        //             "last_price":["0.07300001","BTC"],
+        //             "min_ask":["0.07716895","BTC"],
+        //             "max_bid":["0.0754966","BTC"],
+        //             "volume":["0.168965697","ETH"],
+        //             "price_variation_24h":"-0.046",
+        //             "price_variation_7d":"-0.085"
+        //         }
+        //     }
+        //
         const ticker = this.safeValue (response, 'ticker');
         return this.parseTicker (ticker, market);
     }
 
     parseTicker (ticker, market = undefined) {
+        //
+        // fetchTicker
+        //
+        //     {
+        //         "market_id":"ETH-BTC",
+        //         "last_price":["0.07300001","BTC"],
+        //         "min_ask":["0.07716895","BTC"],
+        //         "max_bid":["0.0754966","BTC"],
+        //         "volume":["0.168965697","ETH"],
+        //         "price_variation_24h":"-0.046",
+        //         "price_variation_7d":"-0.085"
+        //     }
+        //
         const timestamp = this.milliseconds ();
-        let symbol = undefined;
-        if (market !== undefined) {
-            symbol = market['symbol'];
-        }
-        const last = parseFloat (ticker['last_price'][0]);
-        const percentage = parseFloat (ticker['price_variation_24h']);
-        const open = parseFloat (this.priceToPrecision (symbol, last / (percentage + 1)));
-        const change = last - open;
-        const average = this.sum (last, open) / 2;
-        return {
+        const marketId = this.safeString (ticker, 'market_id');
+        const symbol = this.safeSymbol (marketId, market, '-');
+        const lastPrice = this.safeValue (ticker, 'last_price', []);
+        const last = this.safeNumber (lastPrice, 0);
+        const percentage = this.safeNumber (ticker, 'price_variation_24h');
+        return this.safeTicker ({
             'symbol': symbol,
             'timestamp': timestamp,
             'datetime': this.iso8601 (timestamp),
             'high': undefined,
             'low': undefined,
-            'bid': parseFloat (ticker['max_bid'][0]),
+            'bid': this.safeNumber (ticker['max_bid'][0]),
```

Comment:
This is a wrong usage of `safeNumber`, it takes at least 2 arguments – the object to search in, and a key to search for.

This code will break down if there's no `max_bid` and if it's empty, so we should always sanitize the input:

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#accessing-dictionary-keys
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#sanitizing-input-with-safe-methods

A proper code is:

```JavaScript
const maxBid = this.safeValue (ticker, 'max_bid', []);
// ...
    'bid': this.safeNumber (maxBid, 0),
```

The same is true for `ask`.

Reply:
In order to test this, please do the following in fetchTicker:

```
        // const response = await this.publicGetMarketsMarketTicker (this.extend (request, params));
        //
        const response = {
            'ticker': {
                'market_id': 'ETH-BTC',
                'last_price': [ '0.07300001','BTC' ],
                'min_ask': [ '0.07716895','BTC' ],
                'max_bid': ['0.0754966','BTC' ],
                'volume': [ '0.168965697','ETH' ],
                'price_variation_24h': '-0.046',
                'price_variation_7d': '-0.085',
            },
        };
        //
```

Then build and test in all languages

```
npm run build
node examples/js/cli buda fetchTicker --verbose
python3 examples/py/cli.py buda fetchTicker --verbose
php -f examples/php/cli.php buda fetchTicker --verbose
```

If that works, then we can comment it back and merge.